src/annoylib.h

```diff
@@ -88,6 +88,60 @@ inline void normalize(T* v, int f) {
     v[z] /= norm;
 }
 
+template<typename S, typename T, typename Random, typename Distance>
+inline void two_means(const vector<typename Distance::Node*>& nodes, int f, Random& random, bool cosine, T* best_iv, T* best_jv) {
+  size_t count = nodes.size();
+
+  T best_d_sum = numeric_limits<T>::infinity();
+  static vector<T> iv(f, 0), jv(f, 0), iv_sum(f, 0), jv_sum(f, 0);
+  
+  for (int attempt = 0; attempt < 5; attempt++) {
+    size_t i = random.index(count);
+    size_t j = random.index(count-1);
+    j += (j >= i); // ensure that i != j
+    std::copy(&nodes[i]->v[0], &nodes[i]->v[f], &iv[0]);
+    std::copy(&nodes[j]->v[0], &nodes[j]->v[f], &jv[0]);
+    if (cosine) { normalize(&iv[0], f); normalize(&jv[0], f); }
+    T d_sum = 0;
+    
+    for (int iter = 0; iter < 5; iter++) {
+      std::fill(iv_sum.begin(), iv_sum.end(), 0);
+      std::fill(jv_sum.begin(), jv_sum.end(), 0);
+      int ic = 0, jc = 0;
+      for (int l = 0; l < 100 && l < nodes.size(); l++) {
```

Comment:
ah i see, you sample 100 random nodes here, so there could be a bit of variance here...